Handle SVG image subresources better for ImageUtilities callers

annevk opened the pull request:
#### 469ff49e3f3e05e7ea1944d8fe439d476ea87b60
<pre>
Handle SVG image subresources better for ImageUtilities callers
<a href="https://bugs.webkit.org/show_bug.cgi?id=290612">https://bugs.webkit.org/show_bug.cgi?id=290612</a>
<a href="https://rdar.apple.com/147329432">rdar://147329432</a>

Reviewed by Youenn Fablet and Darin Adler.

By waiting for the SVG image&apos;s document to reach the point where it
dispatches the Window load event, we improve handling of typical data:
URL usage.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::~Document):
(WebCore::Document::implicitClose):
(WebCore::Document::whenWindowLoadEventOrDestroyed):
* Source/WebCore/dom/Document.h:
* Source/WebCore/platform/graphics/ImageUtilities.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::tryCreateNativeImageFromData):
(WebCore::createIconDataFromImageData):
(WebCore::decodeImageWithSize):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::subresourcesAreFinished):
(WebCore::SVGImage::tryCreateFromData):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::createIconDataFromImageData):
(WebKit::WebPage::decodeImageData):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/icon-with-subresource.svg: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageData)):
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageDataSVG)):
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageDataSVGWithSubresource)):
* Tools/TestWebKitAPI/cocoa/TestCocoaImageAndCocoaColor.h:
* Tools/TestWebKitAPI/cocoa/TestCocoaImageAndCocoaColor.mm:
(TestWebKitAPI::Util::compareColors):

Canonical link: <a href="https://commits.webkit.org/292944@main">https://commits.webkit.org/292944@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a8a572bea8a853e34351351c759efa5c3a7a8a25

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97551 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17176 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/7392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/102655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/48080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99596 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17469 "Build is in progress. Recent messages:") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/25629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/102655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/48080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100554 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/17469 "Build is in progress. Recent messages:") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/7392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/102655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/17469 "Build is in progress. Recent messages:") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/7392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47522 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/17469 "Build is in progress. Recent messages:") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/7392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104658 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/24631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/25629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/25003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/7392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/82799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/7392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/18224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15756 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/24592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/24414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/27728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/25988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -69,6 +69,30 @@
+    {
+    }
+
+    WeakPtr<LocalDOMWindow, WeakPtrImplWithEventTargetData> m_window;
```
Do we need m_window?

youennf commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -545,14 +569,32 @@ bool isInSVGImage(const Element* element)
+        return;
+    }
+
+    Ref window = *document->protectedWindow();
```
We should probably add a nullptr check and return `completionHandler()` early.
Another way would be to have a Document::whenLoadEventFinished(WTFMove(completionHandler)).

youennf commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -545,14 +569,32 @@ bool isInSVGImage(const Element* element)
+    }
+
+    Ref window = *document->protectedWindow();
+    window->addEventListener(eventNames().loadEvent, SVGImageEventListener::create(window, WTFMove(completionHandler)), { /*capture */ false, /* passive */ std::nullopt, /* once */ true });
```
It is unclear whether the completion handler will be called, say for instance if the window gets removed between the time of adding the event listener and the time the load event can actually fire.
On way of fixing this is to call completion handler if not null in the event listener destructor.

youennf commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -69,6 +69,30 @@
+    }
+
+private:
+    explicit SVGImageEventListener(LocalDOMWindow& window, CompletionHandler<void()>&& completionHandler)
```
No need of explicit with two parameters.

youennf commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -191,13 +191,19 @@ static RefPtr<NativeImage> tryCreateNativeImageFromBitmapImageData(std::span<con
+            completionHandler(nullptr);
+            return;
+        }
+        completionHandler(WTFMove(svgImage->nativeImage(svgImage->size())));
```
```suggestion
        completionHandler(svgImage->nativeImage(svgImage->size()));
```

youennf commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -191,13 +191,19 @@ static RefPtr<NativeImage> tryCreateNativeImageFromBitmapImageData(std::span<con
+        completionHandler(WTFMove(nativeImage));
+        return;
+    }
+    SVGImage::tryCreateFromDataAndWaitForSubresources(data, [completionHandler = WTFMove(completionHandler)] (auto svgImage) mutable {
```
```suggestion
    SVGImage::tryCreateFromDataAndWaitForSubresources(data, [completionHandler = WTFMove(completionHandler)](auto svgImage) mutable {
```

youennf replied:
Here and elsewhere if needed

youennf commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -545,14 +545,29 @@ bool isInSVGImage(const Element* element)
+{
+    ASSERT(rootElement());
+    Ref document = *internalPage()->localTopDocument();
+    if (document->loadEventFinished()) {
```
I would move that logic in `Document::whenWindowLoadEventOrDestroyed`

youennf commented on Source/WebCore/svg/graphics/SVGImage.h:
```diff
@@ -71,6 +71,7 @@ class SVGImage final : public Image {
 
     Page* internalPage() { return m_page.get(); }
     WEBCORE_EXPORT RefPtr<SVGSVGElement> rootElement() const;
+    void subresourcesAreFinished(CompletionHandler<void()>&&);
```
Make it private for now?

youennf commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -4153,7 +4153,7 @@ - (void)_createIconDataFromImageData:(NSData *)imageData withLengths:(NSArray<NS
     }
 
-    auto buffer = WebCore::SharedBuffer::create(imageData);
+    Ref buffer = WebCore::SharedBuffer::create(imageData);
```
Not related to this patch, I would be tempted to not touch this file in this patch.

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
```diff
@@ -312,4 +314,48 @@
+        EXPECT_EQ(result.size.width, 10);
+        EXPECT_EQ(result.size.height, 10);
+
+        // Check two pixel values to see this loaded subresources correctly
```
Nice!

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
```diff
@@ -312,4 +314,48 @@
+        CGImageRef CGImage = [result CGImageForProposedRect:nil context:nil hints:nil];
+        NSBitmapImageRep *bitmap = [[[NSBitmapImageRep alloc] initWithCGImage:CGImage] autorelease];
+
+        // NSLog(@"Color at 3,3: %@", [bitmap colorAtX:3 y: 3]);
```
Do we need this NSLog, ditto below.

darinadler commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
```diff
@@ -312,4 +314,48 @@
     Util::run(&done);
 }
 
+TEST(WebKit, CreateIconDataFromImageDataSVGWithSubresource)
```
Looks like `CGImageForProposedRect` doesn’t exist on iOS so we need to do something else, either make this (temporarily?) a Mac-only test or find a way to get at the pixels without using this Mac-only method.

aprotyas replied:
@annevk you can probably just use `UIImage.CGImage`: https://developer.apple.com/documentation/uikit/uiimage/cgimage?language=objc

annevk replied:
#43305 brings us a little closer, but I still need to understand why I get such a weird value on macOS. And unfortunately the iOS bot didn't print the description field for iOS as I was hoping.

annevk replied:
Okay, I had some lime/green confusion and now the value on macOS seems more reasonable, but it's still a little weird it's not just in sRGB from the start. That's a separate issue though.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -255,41 +261,53 @@ static RefPtr<SharedBuffer> expandSVGImageToData(SVGImage& svgImage, ASCIILitera
 {
     // Supported ICO image sizes by ImageIO.
-    constexpr std::array<unsigned, 5> availableLengths { { 16, 32, 48, 128, 256 } };
+    auto targetLengths = lengths.isEmpty() ? Vector<unsigned>({ 16, 32, 48, 128, 256 }) : lengths;
```
I don’t think we should copy this into a `Vector` until we have to call `SVGImage::tryCreateFromDataAndWaitForSubresources`. We should go back to using `std::span` and `std::array` and then make the vector with `targetLengthVector = Vector<unsigned> { targetLengths }`  instead of `targetLengths = WTFMove(targetLengths)`.

darinadler commented on Source/WebCore/svg/graphics/SVGImage.h:
```diff
@@ -44,7 +44,7 @@ class Settings;
 public:
     static Ref<SVGImage> create(ImageObserver* observer) { return adoptRef(*new SVGImage(observer)); }
-    WEBCORE_EXPORT static RefPtr<SVGImage> tryCreateFromData(std::span<const uint8_t>);
+    WEBCORE_EXPORT static void tryCreateFromDataAndWaitForSubresources(std::span<const uint8_t>, CompletionHandler<void(RefPtr<SVGImage>&&)>&&);
```
Do we need "wait for subresources" in the function name?

annevk replied:
I would prefer keeping this name for now to spread awareness of this edge case as we have numerous places that are not handling this correctly.

darinadler replied:
Why do we need to "spread awareness"? It seems that having only the function that waits for subresources already does the job; hard to do it wrong unless we have a "do it wrong" function. I think the name implies the opposite, that it's reasonable to create an SVG image from data without waiting for subresources.

darinadler replied:
I don't insist on changing the name. Just wanted to make my case for why not to!

annevk replied:
We do have an alternative way of getting an `SVGImage` instance unfortunately. But I'm also persuaded that this could be confusing as well so I'll shorten before landing.

aprotyas commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -836,6 +836,9 @@ Document::~Document()
         clearRareData();
 
+    if (auto completionHandler = std::exchange(m_whenWindowLoadEventOrDestroyed, nullptr))
+        completionHandler();
```
optional: Just calling the completion handler does the exchange with `nullptr` by the way, so this would have been equivalent:

```suggestion
    if (m_whenWindowLoadEventOrDestroyed)
        m_whenWindowLoadEventOrDestroyed();
```

annevk committed:
Handle SVG image subresources better for ImageUtilities callers
https://bugs.webkit.org/show_bug.cgi?id=290612
rdar://147329432

Reviewed by Youenn Fablet and Darin Adler.

By waiting for the SVG image's document to reach the point where it
dispatches the Window load event, we improve handling of typical data:
URL usage.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::~Document):
(WebCore::Document::implicitClose):
(WebCore::Document::whenWindowLoadEventOrDestroyed):
* Source/WebCore/dom/Document.h:
* Source/WebCore/platform/graphics/ImageUtilities.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::tryCreateNativeImageFromData):
(WebCore::createIconDataFromImageData):
(WebCore::decodeImageWithSize):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::subresourcesAreFinished):
(WebCore::SVGImage::tryCreateFromData):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::createIconDataFromImageData):
(WebKit::WebPage::decodeImageData):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/icon-with-subresource.svg: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageData)):
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageDataSVG)):
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageDataSVGWithSubresource)):
* Tools/TestWebKitAPI/cocoa/TestCocoaImageAndCocoaColor.h:
* Tools/TestWebKitAPI/cocoa/TestCocoaImageAndCocoaColor.mm:
(TestWebKitAPI::Util::compareColors):

Canonical link: https://commits.webkit.org/292944@main
